lib/AST/NameLookup.cpp

```diff
@@ -2262,6 +2267,27 @@ void NominalTypeDecl::recordObjCMethod(AbstractFunctionDecl *method,
   vec.push_back(method);
 }
 
+void ClassDecl::recordObjCCategory(ExtensionDecl *ext) {
+  Identifier name = ext->getObjCCategoryName();
+  if (name.empty())
+    return;
+
+  // Use a linear scan on the assumption that there aren't very many extensions.
+  for (auto *other : getExtensions()) {
```

Comment:
I probably don't want the map to be per-`ClassDecl` because the vast majority of classes won't have *any* extensions with a category name, but maybe it can be per-module or something, with a `std::pair<ClassDecl *, Identifier>` key.

Reply:
Request result storage goes in a `DenseMap` unless you choose to customize it to be inline, and you'd only need to allocate the extension map result for classes that have `recordObjCCategory()` called on them so in practice I think it wouldn't create much overhead to do it per-class.

Reply:
The problem is that adding an extension invalidates the lookup table, and `recordCategory(ext)` is called immediately after `ext` is added to its class, so it's likely the table will be incomplete. But if we defer the conflict-checking logic until after we've added all the extensions to the class, we can then use a lookup table to accelerate conflict checking. It's just a bigger redesign of this logic than "drop a lookup table into this existing function", which is what I thought you were suggesting!

Reply:
I see, that makes sense. If you were to delay this check to after all extensions were registered, no extra storage would be needed I guess because it can just be a single pass over the complete list of extensions.

Reply:
Done. While I was working on it, I discovered that some of my request caching code was horribly busted, so fixing that was also nice. :)